File: R/statistics_with_R/05_Exploring_Assumptions/Script_FIles/02_groups_homogeneityOfVariance_normality.R
library(ggplot2)
library(car)
library(pastecs)
library(psych)
library(cowplot)
library(gridExtra)
data<-read.delim('/home/atrides/Desktop/Applied-Statistics-with-R-master/statistics_with_R/05/Data_Files/RExam.dat',header=TRUE)
head(data, 15)


data$uni <-factor(data$uni,levels = c(0:1), labels = c('DunceTown','Sussex'))

is.factor(data$uni)

# exam marks histogram
examplot<-ggplot(data, aes(exam))
examplot<-examplot+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
examplot<-examplot+
  stat_function(fun=dnorm, args=list(mean=mean(data$exam,na.rm = TRUE),sd=sd(data$exam,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')



# computer literacy 
complot<-ggplot(data, aes(computer))
complot<-complot+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
complot<-complot+
  stat_function(fun=dnorm, args=list(mean=mean(data$computer,na.rm = TRUE),sd=sd(data$computer,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')



# lectures attended
lectureplot<-ggplot(data, aes(lectures))
lectureplot<-lectureplot+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
lectureplot<-lectureplot+
  stat_function(fun=dnorm, args=list(mean=mean(data$lectures,na.rm = TRUE),sd=sd(data$lectures,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')



# numeracy
numeracyplot<-ggplot(data, aes(numeracy))
numeracyplot<-numeracyplot+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
numeracyplot<-numeracyplot+
  stat_function(fun=dnorm, args=list(mean=mean(data$numeracy,na.rm = TRUE),sd=sd(data$numeracy,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')
numeracyplot

# using cowplot library, combined multiple plot
plot_grid(examplot, complot, lectureplot, numeracyplot, labels = "AUTO")

# using gridExtra library, compined multiple plots
grid.arrange(examplot, complot,lectureplot,numeracyplot, ncol=2,nrow=2)


stat.desc(data[,cbind('exam', 'computer','lectures','numeracy')], norm=TRUE,basic=FALSE)



# Running the analysis on different groups, i.e as pandas groupby

by(data[,cbind('exam', 'computer', 'lectures','numeracy')], data$uni, describe)

# or also, 
by(data[,cbind('exam', 'computer', 'lectures','numeracy')], data$uni, stat.desc, basic=FALSE, norm=TRUE)

# dividing data as per uni
dunceData<-subset(data, data$uni=='DunceTown')
Sussex<-subset(data, data$uni=='Sussex')


# DunceTown PLot
# exam marks histogram
examplot1<-ggplot(dunceData, aes(exam))
examplot1<-examplot1+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
examplot1<-examplot1+
  stat_function(fun=dnorm, args=list(mean=mean(dunceData$exam,na.rm = TRUE),sd=sd(dunceData$exam,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')


# numeracy
numeracyplot1<-ggplot(dunceData, aes(numeracy))
numeracyplot1<-numeracyplot1+geom_histogram(colour='black',bins=15,fill='white',aes(y=..density..))
numeracyplot1<-numeracyplot1+
  stat_function(fun=dnorm, args=list(mean=mean(dunceData$numeracy,na.rm = TRUE),sd=sd(dunceData$numeracy,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')


# Sussex PLot
# exam marks histogram
examplot2<-ggplot(Sussex, aes(exam))
examplot2<-examplot2+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
examplot2<-examplot2+
  stat_function(fun=dnorm, args=list(mean=mean(Sussex$exam,na.rm = TRUE),sd=sd(Sussex$exam,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')


# numeracy
numeracyplot2<-ggplot(Sussex, aes(numeracy))
numeracyplot2<-numeracyplot2+geom_histogram(colour='black',bins=15,fill='white',aes(y=..density..))
numeracyplot2<-numeracyplot2+
  stat_function(fun=dnorm, args=list(mean=mean(Sussex$numeracy,na.rm = TRUE),sd=sd(Sussex$numeracy,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')

# plotting above things
grid.arrange(examplot1,numeracyplot1,examplot2,numeracyplot2,ncol=2, nrow=2, top='Joint Plots of numeracy and exam score of ( Duncetown vs Sussex )')



# DunceTown PLot
# computer marks histogram
computerplot1<-ggplot(dunceData, aes(computer))
computerplot1<-computerplot1+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
computerplot1<-computerplot1+
  stat_function(fun=dnorm, args=list(mean=mean(dunceData$computer,na.rm = TRUE),sd=sd(dunceData$computer,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')


# lectures
lecplot1<-ggplot(dunceData, aes(lectures))
lecplot1<-lecplot1+geom_histogram(colour='black',bins=15,fill='white',aes(y=..density..))
lecplot1<-lecplot1+
  stat_function(fun=dnorm, args=list(mean=mean(dunceData$lectures,na.rm = TRUE),sd=sd(dunceData$lectures,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')


# Sussex PLot
# exam marks histogram
computerplot2<-ggplot(Sussex, aes(computer))
computerplot2<-computerplot2+geom_histogram(colour='black',bins=30,fill='white',aes(y=..density..))
computerplot2<-computerplot2+
  stat_function(fun=dnorm, args=list(mean=mean(Sussex$computer,na.rm = TRUE),sd=sd(Sussex$computer,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')


# lectures
lecplot2<-ggplot(Sussex, aes(lectures))
lecplot2<-lecplot2+geom_histogram(colour='black',bins=15,fill='white',aes(y=..density..))
lecplot2<-lecplot2+
  stat_function(fun=dnorm, args=list(mean=mean(Sussex$lectures,na.rm = TRUE),sd=sd(Sussex$lectures,na.rm = TRUE)), colour='black', size=1)+
  geom_density(colour='red')

# plotting above things
grid.arrange(computerplot1,lecplot1,computerplot2,lecplot2,ncol=2, nrow=2, top='Joint Plots of numeracy and exam score of ( Duncetown vs Sussex )')



# Doing Statistical tests for normality assumptions

# Shapiro-wilk test whole Data
shapiro.test(data$exam) # non -normal
shapiro.test(data$computer) # normal
shapiro.test(data$lectures) # normal
shapiro.test(data$numeracy) # non-normal


# Shapiro-wilk test Duncetown Data
shapiro.test(dunceData$exam) # normal
shapiro.test(dunceData$computer) # normal
shapiro.test(dunceData$lectures) # normal
shapiro.test(dunceData$numeracy) # non-normal


# Shapiro-wilk test Sussex Data
shapiro.test(Sussex$exam) # normal
shapiro.test(Sussex$computer) # non-normal
shapiro.test(Sussex$lectures) # normal
shapiro.test(Sussex$numeracy) # non-normal


# qqplots
qqplot_exam<- ggplot(data, aes(sample=exam))
qqplot_exam<-qqplot_exam+stat_qq()
qqplot_exam<-qqplot_exam+ggtitle('qqplot Exam')


qqplot_numeracy<- ggplot(data, aes(sample=numeracy))
qqplot_numeracy<-qqplot_numeracy+stat_qq()
qqplot_numeracy<-qqplot_numeracy+ggtitle('qqplot numeracy')

grid.arrange(qqplot_exam,qqplot_numeracy,ncol=2)





# Doing Statistical tests for Homogeneity of variance assumption
# leveneTest() from car package

leveneTest(data$exam, data$uni, center = median)  # assumptions met
leveneTest(data$numeracy, data$uni, center = median)  # assumptions not  met

File: Python/statistics_with_Python/05_Exploring_Assumptions/Script_Files/02_groups_homogeneityOfVariance_normality.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as sp
import numpy as np
import statsmodels.api as sm

university_data = pd.read_csv('/home/atrides/Desktop/R/statistics_with_Python/05_Exploring_Assumptions/Data_Files/RExam.dat', sep='\t')
print(university_data.head())


plt.figure(figsize=(15,12))
x = np.random.normal(university_data['exam'].mean(), university_data['exam'].std(),university_data['exam'].count())
ax1 = plt.subplot(2,2,1)
_ = plt.hist(university_data['exam'],density=True,bins=30, edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(university_data['exam'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,university_data['exam'].max() ])
_ = ax1.set_xlabel('First yr exam score')
_ = ax1.set_ylabel('Density')

x = np.random.normal(university_data['computer'].mean(), university_data['computer'].std(),university_data['computer'].count())
ax2 = plt.subplot(2,2,2)
_ = plt.hist(university_data['computer'],density=True,bins=30, edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(university_data['computer'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,university_data['computer'].max() ])
_ = ax2.set_xlabel('Computer Literacy')
_ = ax2.set_ylabel('Density')

x = np.random.normal(university_data['lectures'].mean(), university_data['lectures'].std(),university_data['lectures'].count())
ax3 = plt.subplot(2,2,3)
_ = plt.hist(university_data['lectures'],density=True,bins=30, edgecolor='black', color='white',linewidth=1.4)
_ = sns.kdeplot(university_data['lectures'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,university_data['lectures'].max() ])
_ = ax3.set_xlabel('Percentage of Lectures Attended')
_ = ax3.set_ylabel('Density')

x = np.random.normal(university_data['numeracy'].mean(), university_data['numeracy'].std(),university_data['numeracy'].count())
ax4 = plt.subplot(2,2,4)
_ = plt.hist(university_data['numeracy'],density=True,bins=30, edgecolor='black', color='white',linewidth=1.4)
_ = sns.kdeplot(university_data['numeracy'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,university_data['numeracy'].max() ])
_ = ax4.set_xlabel('Numeracy')
_ = ax4.set_ylabel('Density')

plt.show()

# Describing basic stats
print(university_data.describe())

# Describing distributional stats
count = 0
for i in university_data.columns[:-1]:
    print('\n'*2,list(university_data.columns)[count], '\n', sp.describe(university_data[i]))
    count+=1


Dunce_data = university_data[university_data['uni']==0]
sussex_data = university_data[university_data['uni']==1]


plt.figure(figsize=(15,12))

x = np.random.normal(Dunce_data['exam'].mean(), Dunce_data['exam'].std(),Dunce_data['exam'].count())
ax1 = plt.subplot(2,2,1)
plt.hist(Dunce_data['exam'],density=True, bins=30,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(Dunce_data['exam'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,Dunce_data['exam'].max() ])
ax1.set_xlabel('First yr exam score')
ax1.set_ylabel('Density')
ax1.set_title('Duncetown University')

x = np.random.normal(sussex_data['exam'].mean(), sussex_data['exam'].std(),sussex_data['exam'].count())
ax2 = plt.subplot(2,2,2,sharey=ax1,sharex=ax1)
plt.hist(sussex_data['exam'],density=True, bins=30,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(sussex_data['exam'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,sussex_data['exam'].max() ])
ax2.set_xlabel('First yr exam score')
ax2.set_ylabel('Density')
ax2.set_title('Sussex University')

x = np.random.normal(Dunce_data['numeracy'].mean(), Dunce_data['numeracy'].std(),Dunce_data['numeracy'].count())
ax3 = plt.subplot(2,2,3)
plt.hist(Dunce_data['numeracy'],density=True, bins=15,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(Dunce_data['numeracy'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,Dunce_data['numeracy'].max() ])
ax3.set_xlabel('Numeracy')
ax3.set_ylabel('Density')

x = np.random.normal(sussex_data['numeracy'].mean(), sussex_data['numeracy'].std(),sussex_data['numeracy'].count())
ax4 = plt.subplot(2,2,4,sharex=ax3, sharey=ax3)
plt.hist(sussex_data['numeracy'],density=True, bins=15,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(sussex_data['numeracy'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,sussex_data['numeracy'].max() ])
ax4.set_xlabel('Numeracy')
ax4.set_ylabel('Density')
               
plt.show()


plt.figure(figsize=(15,12))

x = np.random.normal(Dunce_data['computer'].mean(), Dunce_data['computer'].std(),Dunce_data['computer'].count())
ax1 = plt.subplot(2,2,1)
plt.hist(Dunce_data['computer'],density=True, bins=30,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(Dunce_data['computer'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,Dunce_data['computer'].max() ])
ax1.set_xlabel('computer')
ax1.set_ylabel('Density')
ax1.set_title('Duncetown University')

x = np.random.normal(sussex_data['computer'].mean(), sussex_data['computer'].std(),sussex_data['computer'].count())
ax2 = plt.subplot(2,2,2,sharey=ax1,sharex=ax1)
plt.hist(sussex_data['computer'],density=True, bins=30,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(sussex_data['computer'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,sussex_data['computer'].max() ])
ax2.set_xlabel('computer')
ax2.set_ylabel('Density')
ax2.set_title('Sussex University')

x = np.random.normal(Dunce_data['lectures'].mean(), Dunce_data['lectures'].std(),Dunce_data['lectures'].count())
ax3 = plt.subplot(2,2,3)
plt.hist(Dunce_data['lectures'],density=True, bins=15,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(Dunce_data['lectures'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,Dunce_data['lectures'].max() ])
ax3.set_xlabel('lectures')
ax3.set_ylabel('Density')

x = np.random.normal(sussex_data['lectures'].mean(), sussex_data['lectures'].std(),sussex_data['lectures'].count())
ax4 = plt.subplot(2,2,4,sharex=ax3, sharey=ax3)
plt.hist(sussex_data['lectures'],density=True, bins=15,edgecolor='black',color='white', linewidth=1.4)
_ = sns.kdeplot(sussex_data['lectures'], color='black')
_ = sns.kdeplot(x, color='red')
_ = plt.xlim([0,sussex_data['lectures'].max() ])
ax4.set_xlabel('lectures')
ax4.set_ylabel('Density')
               
plt.show()


# Doing Statistical tests for normality assumptions
# shapiro-wilk test on whole data
print(sp.shapiro(university_data['exam']))      # violates normality
print(sp.shapiro(university_data['computer']))  # normal
print(sp.shapiro(university_data['lectures']))  # normal
print(sp.shapiro(university_data['numeracy']))  # violated normality

# shapiro-wilk test on Duncetown data
print(sp.shapiro(Dunce_data['exam']))      # normal
print(sp.shapiro(Dunce_data['computer']))  # normal
print(sp.shapiro(Dunce_data['lectures']))  # normal
print(sp.shapiro(Dunce_data['numeracy']))  # violated normality

# shapiro-wilk test on Sussex data
print(sp.shapiro(sussex_data['exam']))      # normal
print(sp.shapiro(sussex_data['computer']))  # violated normality
print(sp.shapiro(sussex_data['lectures']))  # normal
print(sp.shapiro(sussex_data['numeracy']))  # violated normality

_ = sm.qqplot(university_data['exam'])  # see deviation from the 45 degree line
plt.show()


_ = sm.qqplot(university_data['numeracy'])   # see deviation from the 45 degree line
plt.show()


# Doing Statistical tests for homogeneity of variance
# levene test for exam score b/w Dunce and sussex¶
print(sp.levene(Dunce_data['exam'], sussex_data['exam']))   # non-significant

# levene test for numeracy b/w Dunce and sussex¶
print(sp.levene(Dunce_data['numeracy'], sussex_data['numeracy']))   # significant deviation

